### DISCUSSION TITLE
fzf single button data source toggling in bash environment variable but in a variable function execution declatation

### DISCUSSION THREAD

**Discussion by MikeLemo1** [junegunn/fzf#3663] [Q&A] [answered]
I've found this fzf snippet that allows fzf to toggle between directory and path's querry and I'm basically trying to make it work for pasting command but I assume all or some of the `[[,=~,(),>,` symbols prevent the script for wotking properly in a variable execution declaration but not sure exacly what needs to be escaped, deleted. changed etc...

Basically I'm trying to make this work, got stuck at the source toggling functionality
```bash
export MY_FZF_OPTS="
  --header 'CTRL-T: Switch between Files/Directories'  
  --bind 'ctrl-d:change-prompt(Directories> )+reload(find * -type d)' 
  --bind 'ctrl-f:change-prompt(Files> )+reload(find * -type f)'
  --bind 'ctrl-y:execute-silent(echo -n {1..} | xclip -se c)'
  --bind 'ctrl-t:transform:[[ ! $FZF_PROMPT =~ Files ]] && echo [change-prompt(Files> )+reload(fd --type file --hidden 2>/dev/null)] || 
  echo [change-prompt(Dirs > )+reload(fd --type directory --hidden 2>/dev/null)] 
  --preview '[[ $FZF_PROMPT =~ Files ]] && bat --color=always {} || tree -C {}'
"
myFzf1() {
  READLINE_LINE_NEW="$(fd --type file --hidden 2>/dev/null | FZF_DEFAULT_OPTS="$MY_FZF_OPTS" $(__fzfcmd) )"
  echo "$MY_FZF_OPTS" # For debugging

  if
    [[ -n $READLINE_LINE_NEW ]]
  then
    builtin bind '"\er": redraw-current-line'
    builtin bind '"\e^": magic-space'
    READLINE_LINE=${READLINE_LINE:+${READLINE_LINE:0:READLINE_POINT}}${READLINE_LINE_NEW}${READLINE_LINE:+${READLINE_LINE:READLINE_POINT}}
    READLINE_POINT=$(( READLINE_POINT + ${#READLINE_LINE_NEW} ))
  else
    builtin bind '"\er":'
    builtin bind '"\e^":'
  fi
}
```

Any idea how to make it work?

**Comment by LangLangBart** [comment, marked as answer]
> echo "$MY_FZF_OPTS" # For debugging

`set -x`[^1][^2] is powerful for debugging as it prints commands and their arguments as they are
executed, allowing you to chip away parts of your code until you find the bug yourself.

---

> Basically I'm trying to make this work, got stuck at the source toggling functionality


In your case, you need to escape the `$FZF_PROMPT` because the surrounding quotes are double quotes.

```bash
# example with surrounding double quotes
export MY_FZF_OPTS="
  --preview '[[ \$FZF_PROMPT =~ Files ]] && bat --color=always {} || tree -C {}'
"

# example with surrounding single quotes
export MY_FZF_OPTS='
  --preview "[[ $FZF_PROMPT =~ Files ]] && bat --color=always {} || tree -C {}"
'
```

Furthermore, your `ctrl-t` construct seems flawed. The `--bind` flag opens with a single quote but
does not end with a single quote. For some reason, you are using `[` and `]` to surround the `echo`
commands, remove them.

Further the parts following the 'echo' command also need to be quoted and need to be escaped[^3].

You should also set a default prompt, because of your `preview` flag.

```bash
export MY_FZF_OPTS='
  --ansi
  --header "CTRL-T: Switch between Files/Directories"
  --bind "ctrl-y:execute-silent(echo -n {} | xclip -se c)"
  --bind "ctrl-t:transform:[[ ! $FZF_PROMPT =~ Files ]] &&
    echo \"change-prompt(Files> )+reload(fd --type file --hidden)\" ||
    echo \"change-prompt(Dirs > )+reload(fd --type directory --hidden)\""
  --preview "[[ $FZF_PROMPT =~ Files ]] && bat --color=always {} || tree -C {}"
  --prompt "Files >"
'
```


[^1]: [Bash Reference Manual - Set Builtin](https://www.gnu.org/software/bash/manual/bash.html#The-Set-Builtin)
[^2]: [bash debugging (wizardzines.com)](https://wizardzines.com/comics/bash-debugging/)
[^3]: [bash quotes (wizardzines.com)](https://wizardzines.com/comics/bash-quotes/)
